{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "7ebd6ed2-b288-4ec0-a1cd-cb7b9f89a31f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pyBigWig\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5edc7a15-0abe-431f-ad00-fa3f81a1ead1",
   "metadata": {},
   "outputs": [],
   "source": [
    "figure3_outpath = '../../../Figures/Figure 3/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5d873a66-8f45-4bd5-9f6f-a4fab80a82ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Target Gene Symbol</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>sgRNA Context Sequence</th>\n",
       "      <th>sgRNA 'Cut' Position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LARS2</td>\n",
       "      <td>chr3</td>\n",
       "      <td>CAGAAAAAAAAAAACAGGACACAGGGGAGA</td>\n",
       "      <td>45389244.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COQ2</td>\n",
       "      <td>chr4</td>\n",
       "      <td>ATAAAAAAAAAAAAGGGCACCAAGTGGCCA</td>\n",
       "      <td>83285687.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>IGBP1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>AATAAAAAAAAAAAGTGATCAGTTTGGAAA</td>\n",
       "      <td>70133000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>YEATS4</td>\n",
       "      <td>chr12</td>\n",
       "      <td>AAAAAAAAAAAAAATTAACGTGCCAGGGGC</td>\n",
       "      <td>69360661.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>COPS3</td>\n",
       "      <td>chr17</td>\n",
       "      <td>AAAAAAAAAAAAAGCCGTGTGAGCTGGCGG</td>\n",
       "      <td>17280422.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257411</th>\n",
       "      <td>WDR11</td>\n",
       "      <td>chr10</td>\n",
       "      <td>GGCTAGGACCTCTACTACAATGAATGGGAG</td>\n",
       "      <td>120855890.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257412</th>\n",
       "      <td>WDR11</td>\n",
       "      <td>chr10</td>\n",
       "      <td>GCTAGGACCTCTACTACAATGAATGGGAGA</td>\n",
       "      <td>120855891.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257413</th>\n",
       "      <td>WDR11</td>\n",
       "      <td>chr10</td>\n",
       "      <td>CAGTCTTTCATCAAGTCTGATGTAAGGTAT</td>\n",
       "      <td>120855971.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257414</th>\n",
       "      <td>WDR11</td>\n",
       "      <td>chr10</td>\n",
       "      <td>TTTATGTAGGTCCAAGTTTCAGTCTGGTAT</td>\n",
       "      <td>120856094.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257415</th>\n",
       "      <td>WDR11</td>\n",
       "      <td>chr10</td>\n",
       "      <td>GTTCTTAAAGTGCCTGTCTAATATTGGTTG</td>\n",
       "      <td>120856156.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>122029 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Target Gene Symbol chromosome          sgRNA Context Sequence  \\\n",
       "0                   LARS2       chr3  CAGAAAAAAAAAAACAGGACACAGGGGAGA   \n",
       "3                    COQ2       chr4  ATAAAAAAAAAAAAGGGCACCAAGTGGCCA   \n",
       "6                   IGBP1       chrX  AATAAAAAAAAAAAGTGATCAGTTTGGAAA   \n",
       "9                  YEATS4      chr12  AAAAAAAAAAAAAATTAACGTGCCAGGGGC   \n",
       "12                  COPS3      chr17  AAAAAAAAAAAAAGCCGTGTGAGCTGGCGG   \n",
       "...                   ...        ...                             ...   \n",
       "257411              WDR11      chr10  GGCTAGGACCTCTACTACAATGAATGGGAG   \n",
       "257412              WDR11      chr10  GCTAGGACCTCTACTACAATGAATGGGAGA   \n",
       "257413              WDR11      chr10  CAGTCTTTCATCAAGTCTGATGTAAGGTAT   \n",
       "257414              WDR11      chr10  TTTATGTAGGTCCAAGTTTCAGTCTGGTAT   \n",
       "257415              WDR11      chr10  GTTCTTAAAGTGCCTGTCTAATATTGGTTG   \n",
       "\n",
       "        sgRNA 'Cut' Position  \n",
       "0                 45389244.0  \n",
       "3                 83285687.0  \n",
       "6                 70133000.0  \n",
       "9                 69360661.0  \n",
       "12                17280422.0  \n",
       "...                      ...  \n",
       "257411           120855890.0  \n",
       "257412           120855891.0  \n",
       "257413           120855971.0  \n",
       "257414           120856094.0  \n",
       "257415           120856156.0  \n",
       "\n",
       "[122029 rows x 4 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read in cleaned combined data\n",
    "NewDatasetCombine = pd.read_csv('../../../Data/cleanedData/CombinedCleanedDatasets.csv', low_memory=False)\n",
    "\n",
    "#get dataframe of unique guide/gene combinations across datasets\n",
    "UniqueNewDatasetCombine = NewDatasetCombine[['Target Gene Symbol',\n",
    "                                             'chromosome',\n",
    "                                             'sgRNA Context Sequence',\n",
    "                                             'sgRNA \\'Cut\\' Position']].drop_duplicates()\n",
    "\n",
    "#alter chromosome notation to be compatible with bed files\n",
    "UniqueNewDatasetCombine['chromosome'] = UniqueNewDatasetCombine['chromosome'].replace('chr24','chrY')\n",
    "UniqueNewDatasetCombine['chromosome'] = UniqueNewDatasetCombine['chromosome'].replace('chr23','chrX')\n",
    "UniqueNewDatasetCombine"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "118f7d75-106b-418f-9a0e-1ce96ecff592",
   "metadata": {},
   "source": [
    "### Download ATAC-seq files\n",
    "\n",
    "Download bigBed files containing pseudo replicated peaks derived from ATAC-seq data:\n",
    "\n",
    "- A549s: https://www.encodeproject.org/files/ENCFF808YMI/\n",
    "- HCT116s: https://www.encodeproject.org/files/ENCFF068PVP/\n",
    "- K562s: https://www.encodeproject.org/files/ENCFF086JCJ/\n",
    "- HepG2s: https://www.encodeproject.org/files/ENCFF906NBO/\n",
    "- GM12878s: https://www.encodeproject.org/files/ENCFF285BIH/\n",
    "- Panc1s: https://www.encodeproject.org/files/ENCFF882XTL/\n",
    "- MCF-7s: https://www.encodeproject.org/files/ENCFF502RTI/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e2d1ed20-a4fd-4713-a0b8-cd773a281b98",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in ATAC bigBed files\n",
    "file_path = '../../../Data/ATAC-seq/raw/'\n",
    "atac_a549_bb = pyBigWig.open(file_path+'ENCFF808YMI.bigBed')\n",
    "atac_hct116_bb = pyBigWig.open(file_path+'ENCFF068PVP.bigBed')\n",
    "atac_k562_bb = pyBigWig.open(file_path+'ENCFF086JCJ.bigBed')\n",
    "atac_Panc1_bb = pyBigWig.open(file_path+'ENCFF882XTL.bigBed')\n",
    "atac_MCF7_bb = pyBigWig.open(file_path+'ENCFF502RTI.bigBed')\n",
    "atac_GM12878_bb = pyBigWig.open(file_path+'ENCFF285BIH.bigBed')\n",
    "atac_HepG2_bb = pyBigWig.open(file_path+'ENCFF906NBO.bigBed')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58d9ed2e-94bf-48d5-809e-d26fb0fdf344",
   "metadata": {},
   "source": [
    "### Compute Peak Overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b651b286-fbbf-46d2-b112-484dbebbe34a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def isInRange(n, tbl):\n",
    "    #https://stackoverflow.com/questions/9019581/what-does-numpy-apply-along-axis-perform-exactly\n",
    "    #https://stackoverflow.com/questions/71377816/check-if-value-is-between-two-values-in-numpy-array\n",
    "    # n: is sgRNA location in the tiling/phenotypic dataset (need to subtract 1 because CRISPick is 1-based and bed files are 0-based)\n",
    "    # tbl: numpy file of dataframe of bigbed with chro, start, and end of peak filtered_peak_file.to_numpy()\n",
    "    #apply_along_axis applies the supplied function along 1D slices of the input array, \n",
    "    #with the slices taken along the axis you specify. \n",
    "    return sum(np.apply_along_axis(lambda row: row[0] <= n-1 < row[1], 1, tbl)\\\n",
    "        .tolist())\n",
    "\n",
    "def peak_overlap(actual_tiling , bigbed_df, chrom, sgrna_location_col, chromosome_col = 'chromosome', \n",
    "                 gene_col = 'gene symbol'):\n",
    "    # This function find whether guide overlaps with a peak in one chromosome \n",
    "    \n",
    "    # actual_tiling: CRISPRi data (or any CRISPR data) with at least three columns \n",
    "    #                       [a location to indicate sgRNA, Gene name, Chromosome]\n",
    "    #                       Note this DOES NOT require unique sgRNA location \n",
    "    # bigbed_df: ATAC seq from ENCODE in bigBed format and read through pyBigWig\n",
    "    # chrom: string of chromosome such as 'chr1'\n",
    "    # sgrna_location_col: string of column for sgRNA location\n",
    "    \n",
    "    # returns a dataframe, same as the actual_tilling but with an additional column overlap with peak to indicate \n",
    "    # whether theres an overlap between start and end of a peak and sgRNA location for one chromosome\n",
    "    \n",
    "    #----------------------------------------------------------------------------------------------------\n",
    "    print(chrom)\n",
    "    #subset chromsome number\n",
    "    tiling_subset_chromo =  actual_tiling[actual_tiling[chromosome_col] == chrom].copy(deep=True)\n",
    "    \n",
    "    # select unique pam coord and chr and gene - remove duplicate \n",
    "    # tiling_lib: CRISPRi data (or any CRISPR data) with at least three columns \n",
    "    #                       [a location to indicate sgRNA, Gene name, Chromosome]\n",
    "    #                       Note this should has unique sgRNA location \n",
    "    tiling_lib = tiling_subset_chromo.drop_duplicates(subset=[sgrna_location_col, chromosome_col,gene_col]).copy(deep=True)\n",
    "    \n",
    "    \n",
    "    # change from string to int\n",
    "    tiling_lib[sgrna_location_col] = list(map(int,tiling_lib[sgrna_location_col]))\n",
    "    \n",
    "    #obtain the smallest pam coord in a specific chromosome\n",
    "    smallest_pam_coord = tiling_lib[sgrna_location_col].min()\n",
    "    #obtain the largest pam coord in a specific chromosome\n",
    "    largestest_pam_coord = tiling_lib[sgrna_location_col].max()\n",
    "\n",
    "    \n",
    "    #subset chrom number and having the end coord to be larger than the smallest pam coord\n",
    "    \n",
    "    # Retrieving bigBed entries in https://github.com/deeptools/pyBigWig explains\n",
    "    # filtered_peak_file returns a list of tuple of (Start position in chromosome, End position in chromosome)\n",
    "    filtered_peak_file = bigbed_df.entries(chrom, smallest_pam_coord, largestest_pam_coord, withString=False) \n",
    "    \n",
    "    new_tiling_lib = tiling_lib.copy(deep=True)\n",
    "\n",
    "    # make sure the selected peak file is not none\n",
    "    if filtered_peak_file is not None:\n",
    "    \n",
    "        # only kept unique peaks and make it into dataframe because of https://www.biostars.org/p/464618/\n",
    "        filtered_peak_file = pd.DataFrame(set(filtered_peak_file))\n",
    "\n",
    "\n",
    "        # iterating over every single sgRNA location in the tiling library/a dataset with phenotypic data\n",
    "        peak_list = [isInRange(x, filtered_peak_file.to_numpy()) \n",
    "                     for x in tqdm(np.nditer(tiling_lib[sgrna_location_col].to_numpy()), \n",
    "                                   total=len(tiling_lib[sgrna_location_col]),\n",
    "                                   desc='number of unique sgRNA position and gene symbol for ' + chrom)]\n",
    "    # if its None then return a list of 0 to show there is no overlaps\n",
    "    else:\n",
    "        peak_list = [0] *len(new_tiling_lib)\n",
    "\n",
    "    \n",
    "    #https://stackoverflow.com/questions/32573452/settingwithcopywarning-even-when-using-locrow-indexer-col-indexer-value\n",
    "    new_tiling_lib.loc[:, 'overlap with peak'] = peak_list\n",
    "    \n",
    "    new_tiling_lib = new_tiling_lib[[gene_col, chromosome_col, sgrna_location_col, 'overlap with peak']]\n",
    "    #b['overlap with peak'] = peak_list\n",
    "    chr_df = pd.merge(actual_tiling, new_tiling_lib, on = [gene_col, chromosome_col, sgrna_location_col])\n",
    "\n",
    "    return chr_df\n",
    "\n",
    "def ATACseq_run(actual_tiling , bigbed_df, sgrna_location_col, chromosome_col = 'chromosome', gene_col = 'gene symbol'):\n",
    "    anno_peak = pd.DataFrame()\n",
    "    \n",
    "    # This function iterate over all the chromosome to run the function [peak_overlap] and concatenate them together\n",
    "    # actual_tiling: CRISPRi data (or any CRISPR data) with at least three columns \n",
    "    #                       [a location to indicate sgRNA, Gene name, Chromosome]\n",
    "    #                       Note this DOES NOT require unique sgRNA location \n",
    "    # bigbed_df: ATAC seq from ENCODE in bigBed format and read through pyBigWig\n",
    "    # sgrna_location_col: string of column name for sgRNA location\n",
    "    # chromosome_col: string of column name for chromosome\n",
    "    # gene_col: string of column name for 'gene symbol'\n",
    "    \n",
    "    # returns a dataframe, same as the actual_tilling but with an additional column overlap with peak to indicate \n",
    "    # whether theres an overlap between start and end of a peak and sgRNA location for mutiple chromosomes\n",
    "    \n",
    "    # a list of chromosome in the data\n",
    "    chromo_list = set(actual_tiling[chromosome_col])\n",
    "    print('There are total', len(chromo_list), \n",
    "      ' chromosome in the dataset and they are ',set(chromo_list)) \n",
    "    \n",
    "    #iterate peak_overlap function for all the chromosome\n",
    "    for chromo in chromo_list:\n",
    "        print(chromo)\n",
    "        chrom_num_df = peak_overlap(actual_tiling = actual_tiling, bigbed_df = bigbed_df, chrom = chromo, \n",
    "                                    sgrna_location_col = sgrna_location_col, \n",
    "                                    chromosome_col = chromosome_col, gene_col = gene_col)\n",
    "        anno_peak = pd.concat([anno_peak, chrom_num_df])\n",
    "        \n",
    "    return anno_peak"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "cba3929c-b5af-4c51-af86-70245f52d370",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute peak overlap in each cell line\n",
    "\n",
    "# chrtest_a549 = ATACseq_run(actual_tiling= UniqueNewDatasetCombine, bigbed_df = atac_a549_bb, sgrna_location_col = 'sgRNA \\'Cut\\' Position',\n",
    "#                            chromosome_col = 'chromosome', gene_col = 'Target Gene Symbol' )\n",
    "\n",
    "# chrtest_hct116 = ATACseq_run(actual_tiling= UniqueNewDatasetCombine, bigbed_df = atac_hct116_bb, sgrna_location_col = 'sgRNA \\'Cut\\' Position',\n",
    "#                              chromosome_col = 'chromosome', gene_col = 'Target Gene Symbol' )\n",
    "\n",
    "# chrtest_k562 = ATACseq_run(actual_tiling= UniqueNewDatasetCombine, bigbed_df = atac_k562_bb, sgrna_location_col = 'sgRNA \\'Cut\\' Position',\n",
    "#                            chromosome_col = 'chromosome', gene_col = 'Target Gene Symbol' )\n",
    "\n",
    "# chrtest_Panc1 = ATACseq_run(actual_tiling= UniqueNewDatasetCombine, bigbed_df = atac_Panc1_bb, sgrna_location_col = 'sgRNA \\'Cut\\' Position',\n",
    "#                             chromosome_col = 'chromosome', gene_col = 'Target Gene Symbol' )\n",
    "\n",
    "# chrtest_MCF7 = ATACseq_run(actual_tiling= UniqueNewDatasetCombine, bigbed_df = atac_MCF7_bb, sgrna_location_col = 'sgRNA \\'Cut\\' Position',\n",
    "#                            chromosome_col = 'chromosome', gene_col = 'Target Gene Symbol' )\n",
    "\n",
    "# chrtest_GM12878 = ATACseq_run(actual_tiling= UniqueNewDatasetCombine, bigbed_df = atac_GM12878_bb, sgrna_location_col = 'sgRNA \\'Cut\\' Position',\n",
    "#                               chromosome_col = 'chromosome', gene_col = 'Target Gene Symbol' )\n",
    "\n",
    "# chrtest_HepG2 = ATACseq_run(actual_tiling= UniqueNewDatasetCombine, bigbed_df = atac_HepG2_bb, sgrna_location_col = 'sgRNA \\'Cut\\' Position',\n",
    "#                             chromosome_col = 'chromosome', gene_col = 'Target Gene Symbol' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "59f727ed-d581-4518-adb8-00892851958c",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrtest_a549.to_csv('../../../Data/ATAC-seq/peakOverlap/A549.csv', index = False)\n",
    "chrtest_hct116.to_csv('../../../Data/ATAC-seq/peakOverlap/HCT116.csv', index = False)\n",
    "chrtest_k562.to_csv('../../../Data/ATAC-seq/peakOverlap/K562.csv', index = False)\n",
    "chrtest_Panc1.to_csv('../../../Data/ATAC-seq/peakOverlap/Panc1.csv', index = False)\n",
    "chrtest_MCF7.to_csv('../../../Data/ATAC-seq/peakOverlap/MCF7.csv', index = False)\n",
    "chrtest_GM12878.to_csv('../../../Data/ATAC-seq/peakOverlap/GM12878.csv', index = False)\n",
    "chrtest_HepG2.to_csv('../../../Data/ATAC-seq/peakOverlap/HepG2.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eed965d8-df00-4d07-bd62-b4c77ff4d08c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "crispri",
   "language": "python",
   "name": "crispri"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
