{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b96dac06-03f1-47dc-a248-de8798509d64",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from zipfile import ZipFile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "21b66db4-ed44-4bee-946e-34c04e79eeab",
   "metadata": {},
   "outputs": [],
   "source": [
    "zip_file = ZipFile('../../Data/modelTraining/trainData.csv.zip')\n",
    "train_df = pd.read_csv(zip_file.open('trainData.csv'))\n",
    "test_df = pd.read_csv('../../Data/modelTraining/testData.csv')\n",
    "cleaned_data = pd.read_csv('../../Data/cleanedData/CleanedScaledData.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1134c85e-1991-4e69-a22d-bcb3f810b6ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Target Gene Symbol</th>\n",
       "      <th>sgRNA Context Sequence</th>\n",
       "      <th>Domain</th>\n",
       "      <th>Strand of Target</th>\n",
       "      <th>Strand of sgRNA</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>sgRNA 'Cut' Position</th>\n",
       "      <th>TSS Position</th>\n",
       "      <th>sgRNA 'Cut' Site TSS Offset</th>\n",
       "      <th>rs3ChenSeqScore</th>\n",
       "      <th>doubleZscore</th>\n",
       "      <th>DHS overlap with peak proportion</th>\n",
       "      <th>ATAC overlap with peak proportion</th>\n",
       "      <th>H3K4me3</th>\n",
       "      <th>H3k27ac</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>TSS Bucket (-5kb to 5kb)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AATF</td>\n",
       "      <td>AAAACTAAGTTGGAGACTGATGGAAGGATT</td>\n",
       "      <td>Kox1</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr17</td>\n",
       "      <td>36949636.0</td>\n",
       "      <td>36948955.0</td>\n",
       "      <td>681.0</td>\n",
       "      <td>0.517537</td>\n",
       "      <td>-0.169142</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>++</td>\n",
       "      <td>[675, 700)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AATF</td>\n",
       "      <td>AAAACTAAGTTGGAGACTGATGGAAGGATT</td>\n",
       "      <td>Zim3</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr17</td>\n",
       "      <td>36949636.0</td>\n",
       "      <td>36948955.0</td>\n",
       "      <td>681.0</td>\n",
       "      <td>0.517537</td>\n",
       "      <td>-0.340682</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>++</td>\n",
       "      <td>[675, 700)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AATF</td>\n",
       "      <td>AAACTTGGTGCTCGGCTGGATGGGCGGAAC</td>\n",
       "      <td>Kox1</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr17</td>\n",
       "      <td>36949430.0</td>\n",
       "      <td>36948955.0</td>\n",
       "      <td>475.0</td>\n",
       "      <td>0.011229</td>\n",
       "      <td>0.335486</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>++</td>\n",
       "      <td>[475, 500)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AATF</td>\n",
       "      <td>AAACTTGGTGCTCGGCTGGATGGGCGGAAC</td>\n",
       "      <td>Zim3</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr17</td>\n",
       "      <td>36949430.0</td>\n",
       "      <td>36948955.0</td>\n",
       "      <td>475.0</td>\n",
       "      <td>0.011229</td>\n",
       "      <td>0.480443</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>++</td>\n",
       "      <td>[475, 500)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AATF</td>\n",
       "      <td>AAAGAAGCCGAAGCCCTCCTCCCGAGGCCG</td>\n",
       "      <td>Kox1</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17</td>\n",
       "      <td>36949375.0</td>\n",
       "      <td>36948955.0</td>\n",
       "      <td>420.0</td>\n",
       "      <td>0.316988</td>\n",
       "      <td>1.393357</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>+-</td>\n",
       "      <td>[400, 425)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163529</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>TTTAATGGTTAACACAGATTAGGGTGGGTT</td>\n",
       "      <td>Kox1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>85708727.0</td>\n",
       "      <td>85708433.0</td>\n",
       "      <td>-293.0</td>\n",
       "      <td>0.611774</td>\n",
       "      <td>-0.458460</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>--</td>\n",
       "      <td>[-300, -275)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163530</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>TTTACGGCTCTGCGGAGGCCCTGCCGGATT</td>\n",
       "      <td>Kox1</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>85708514.0</td>\n",
       "      <td>85708433.0</td>\n",
       "      <td>-80.0</td>\n",
       "      <td>-0.431003</td>\n",
       "      <td>-0.478862</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-+</td>\n",
       "      <td>[-100, -75)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163531</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>TTTAGTGGTAAAAGAAGCGAAGGTGGGTGA</td>\n",
       "      <td>Kox1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>85707893.0</td>\n",
       "      <td>85708433.0</td>\n",
       "      <td>541.0</td>\n",
       "      <td>0.557163</td>\n",
       "      <td>-0.386223</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>--</td>\n",
       "      <td>[525, 550)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163532</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>TTTCTGGAGTCAGACGGCGTTCGCGGGGGT</td>\n",
       "      <td>Kox1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>85707491.0</td>\n",
       "      <td>85708433.0</td>\n",
       "      <td>943.0</td>\n",
       "      <td>-0.492183</td>\n",
       "      <td>-0.296669</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>--</td>\n",
       "      <td>[925, 950)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163533</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>TTTGTGGCTGCTTGGCTCCACACGTGGGCC</td>\n",
       "      <td>Kox1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>85708387.0</td>\n",
       "      <td>85708433.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>-0.326489</td>\n",
       "      <td>-0.629514</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>--</td>\n",
       "      <td>[25, 50)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127517 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Target Gene Symbol          sgRNA Context Sequence Domain  \\\n",
       "0                    AATF  AAAACTAAGTTGGAGACTGATGGAAGGATT   Kox1   \n",
       "1                    AATF  AAAACTAAGTTGGAGACTGATGGAAGGATT   Zim3   \n",
       "2                    AATF  AAACTTGGTGCTCGGCTGGATGGGCGGAAC   Kox1   \n",
       "3                    AATF  AAACTTGGTGCTCGGCTGGATGGGCGGAAC   Zim3   \n",
       "4                    AATF  AAAGAAGCCGAAGCCCTCCTCCCGAGGCCG   Kox1   \n",
       "...                   ...                             ...    ...   \n",
       "163529             ZNHIT6  TTTAATGGTTAACACAGATTAGGGTGGGTT   Kox1   \n",
       "163530             ZNHIT6  TTTACGGCTCTGCGGAGGCCCTGCCGGATT   Kox1   \n",
       "163531             ZNHIT6  TTTAGTGGTAAAAGAAGCGAAGGTGGGTGA   Kox1   \n",
       "163532             ZNHIT6  TTTCTGGAGTCAGACGGCGTTCGCGGGGGT   Kox1   \n",
       "163533             ZNHIT6  TTTGTGGCTGCTTGGCTCCACACGTGGGCC   Kox1   \n",
       "\n",
       "       Strand of Target Strand of sgRNA chromosome  sgRNA 'Cut' Position  \\\n",
       "0                     +               +      chr17            36949636.0   \n",
       "1                     +               +      chr17            36949636.0   \n",
       "2                     +               +      chr17            36949430.0   \n",
       "3                     +               +      chr17            36949430.0   \n",
       "4                     +               -      chr17            36949375.0   \n",
       "...                 ...             ...        ...                   ...   \n",
       "163529                -               -       chr1            85708727.0   \n",
       "163530                -               +       chr1            85708514.0   \n",
       "163531                -               -       chr1            85707893.0   \n",
       "163532                -               -       chr1            85707491.0   \n",
       "163533                -               -       chr1            85708387.0   \n",
       "\n",
       "        TSS Position  sgRNA 'Cut' Site TSS Offset  rs3ChenSeqScore  \\\n",
       "0         36948955.0                        681.0         0.517537   \n",
       "1         36948955.0                        681.0         0.517537   \n",
       "2         36948955.0                        475.0         0.011229   \n",
       "3         36948955.0                        475.0         0.011229   \n",
       "4         36948955.0                        420.0         0.316988   \n",
       "...              ...                          ...              ...   \n",
       "163529    85708433.0                       -293.0         0.611774   \n",
       "163530    85708433.0                        -80.0        -0.431003   \n",
       "163531    85708433.0                        541.0         0.557163   \n",
       "163532    85708433.0                        943.0        -0.492183   \n",
       "163533    85708433.0                         47.0        -0.326489   \n",
       "\n",
       "        doubleZscore  DHS overlap with peak proportion  \\\n",
       "0          -0.169142                               0.0   \n",
       "1          -0.340682                               0.0   \n",
       "2           0.335486                               0.0   \n",
       "3           0.480443                               0.0   \n",
       "4           1.393357                               0.0   \n",
       "...              ...                               ...   \n",
       "163529     -0.458460                               0.0   \n",
       "163530     -0.478862                               1.0   \n",
       "163531     -0.386223                               0.0   \n",
       "163532     -0.296669                               0.0   \n",
       "163533     -0.629514                               1.0   \n",
       "\n",
       "        ATAC overlap with peak proportion  H3K4me3   H3k27ac Orientation  \\\n",
       "0                                0.285714      1.0  1.000000          ++   \n",
       "1                                0.285714      1.0  1.000000          ++   \n",
       "2                                1.000000      1.0  1.000000          ++   \n",
       "3                                1.000000      1.0  1.000000          ++   \n",
       "4                                1.000000      1.0  1.000000          +-   \n",
       "...                                   ...      ...       ...         ...   \n",
       "163529                           1.000000      1.0  1.000000          --   \n",
       "163530                           1.000000      1.0  1.000000          -+   \n",
       "163531                           0.857143      1.0  1.000000          --   \n",
       "163532                           0.142857      1.0  0.666667          --   \n",
       "163533                           1.000000      1.0  1.000000          --   \n",
       "\n",
       "       TSS Bucket (-5kb to 5kb)  \n",
       "0                    [675, 700)  \n",
       "1                    [675, 700)  \n",
       "2                    [475, 500)  \n",
       "3                    [475, 500)  \n",
       "4                    [400, 425)  \n",
       "...                         ...  \n",
       "163529             [-300, -275)  \n",
       "163530              [-100, -75)  \n",
       "163531               [525, 550)  \n",
       "163532               [925, 950)  \n",
       "163533                 [25, 50)  \n",
       "\n",
       "[127517 rows x 17 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df_annotated = cleaned_data[cleaned_data['Target Gene Symbol'].isin(train_df['Target_Gene_Symbol'])]\n",
    "test_df_annotated = cleaned_data[cleaned_data['Target Gene Symbol'].isin(test_df['Target_Gene_Symbol'])]\n",
    "train_df_annotated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "363c8077-cad9-47a4-8072-e094d8d60db9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Target Gene Symbol</th>\n",
       "      <th>sgRNA 'Cut' Site TSS Offset</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>Strand of Target</th>\n",
       "      <th>sgRNA 'Cut' Position</th>\n",
       "      <th>sgRNA Context Sequence</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>rs3ChenSeqScore</th>\n",
       "      <th>ATAC overlap with peak proportion</th>\n",
       "      <th>doubleZscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AATF</td>\n",
       "      <td>-1000.0</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>36947955.0</td>\n",
       "      <td>GTTTAGAGGGGTCTAGGTCTCTACAGGCAA</td>\n",
       "      <td>++</td>\n",
       "      <td>0.782222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.055701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AATF</td>\n",
       "      <td>-985.0</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>36947970.0</td>\n",
       "      <td>GTCTATTCTTATGGTATCTGTTGGAGGTTG</td>\n",
       "      <td>+-</td>\n",
       "      <td>-0.267594</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.080429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AATF</td>\n",
       "      <td>-982.0</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>36947973.0</td>\n",
       "      <td>GTGGTCTATTCTTATGGTATCTGTTGGAGG</td>\n",
       "      <td>+-</td>\n",
       "      <td>0.110804</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.836591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AATF</td>\n",
       "      <td>-972.0</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>36947983.0</td>\n",
       "      <td>TGACTACAAAGTGGTCTATTCTTATGGTAT</td>\n",
       "      <td>+-</td>\n",
       "      <td>-0.103015</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.696359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AATF</td>\n",
       "      <td>-959.0</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>36947996.0</td>\n",
       "      <td>GGGATGATAAGTGTGACTACAAAGTGGTCT</td>\n",
       "      <td>+-</td>\n",
       "      <td>0.506514</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.085971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95313</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>948.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>85707486.0</td>\n",
       "      <td>GGAGTCAGACGGCGTTCGCGGGGGTGGGTT</td>\n",
       "      <td>--</td>\n",
       "      <td>-1.255058</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>-0.840383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95314</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>949.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>85707485.0</td>\n",
       "      <td>GAGTCAGACGGCGTTCGCGGGGGTGGGTTG</td>\n",
       "      <td>--</td>\n",
       "      <td>-0.381792</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>-0.637729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95315</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>1025.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>85707409.0</td>\n",
       "      <td>ATACATGAAATTAGTAGTGTCCAATGGGTA</td>\n",
       "      <td>--</td>\n",
       "      <td>-0.244887</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.527092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95316</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>1026.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>85707408.0</td>\n",
       "      <td>TACATGAAATTAGTAGTGTCCAATGGGTAA</td>\n",
       "      <td>--</td>\n",
       "      <td>0.080581</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.173052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95317</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>1032.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>85707402.0</td>\n",
       "      <td>AAATTAGTAGTGTCCAATGGGTAATGGTCA</td>\n",
       "      <td>--</td>\n",
       "      <td>0.326828</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.626674</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>95318 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Target Gene Symbol  sgRNA 'Cut' Site TSS Offset chromosome  \\\n",
       "0                   AATF                      -1000.0      chr17   \n",
       "1                   AATF                       -985.0      chr17   \n",
       "2                   AATF                       -982.0      chr17   \n",
       "3                   AATF                       -972.0      chr17   \n",
       "4                   AATF                       -959.0      chr17   \n",
       "...                  ...                          ...        ...   \n",
       "95313             ZNHIT6                        948.0       chr1   \n",
       "95314             ZNHIT6                        949.0       chr1   \n",
       "95315             ZNHIT6                       1025.0       chr1   \n",
       "95316             ZNHIT6                       1026.0       chr1   \n",
       "95317             ZNHIT6                       1032.0       chr1   \n",
       "\n",
       "      Strand of Target  sgRNA 'Cut' Position          sgRNA Context Sequence  \\\n",
       "0                    +            36947955.0  GTTTAGAGGGGTCTAGGTCTCTACAGGCAA   \n",
       "1                    +            36947970.0  GTCTATTCTTATGGTATCTGTTGGAGGTTG   \n",
       "2                    +            36947973.0  GTGGTCTATTCTTATGGTATCTGTTGGAGG   \n",
       "3                    +            36947983.0  TGACTACAAAGTGGTCTATTCTTATGGTAT   \n",
       "4                    +            36947996.0  GGGATGATAAGTGTGACTACAAAGTGGTCT   \n",
       "...                ...                   ...                             ...   \n",
       "95313                -            85707486.0  GGAGTCAGACGGCGTTCGCGGGGGTGGGTT   \n",
       "95314                -            85707485.0  GAGTCAGACGGCGTTCGCGGGGGTGGGTTG   \n",
       "95315                -            85707409.0  ATACATGAAATTAGTAGTGTCCAATGGGTA   \n",
       "95316                -            85707408.0  TACATGAAATTAGTAGTGTCCAATGGGTAA   \n",
       "95317                -            85707402.0  AAATTAGTAGTGTCCAATGGGTAATGGTCA   \n",
       "\n",
       "      Orientation  rs3ChenSeqScore  ATAC overlap with peak proportion  \\\n",
       "0              ++         0.782222                           0.000000   \n",
       "1              +-        -0.267594                           0.000000   \n",
       "2              +-         0.110804                           0.000000   \n",
       "3              +-        -0.103015                           0.000000   \n",
       "4              +-         0.506514                           0.000000   \n",
       "...           ...              ...                                ...   \n",
       "95313          --        -1.255058                           0.142857   \n",
       "95314          --        -0.381792                           0.142857   \n",
       "95315          --        -0.244887                           0.000000   \n",
       "95316          --         0.080581                           0.000000   \n",
       "95317          --         0.326828                           0.000000   \n",
       "\n",
       "       doubleZscore  \n",
       "0         -1.055701  \n",
       "1         -1.080429  \n",
       "2         -0.836591  \n",
       "3         -0.696359  \n",
       "4          1.085971  \n",
       "...             ...  \n",
       "95313     -0.840383  \n",
       "95314     -0.637729  \n",
       "95315     -0.527092  \n",
       "95316     -0.173052  \n",
       "95317     -0.626674  \n",
       "\n",
       "[95318 rows x 10 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df_unique = train_df_annotated.groupby(['Target Gene Symbol','sgRNA \\'Cut\\' Site TSS Offset','chromosome','Strand of Target',\n",
    "                                              'sgRNA \\'Cut\\' Position','sgRNA Context Sequence','Orientation','rs3ChenSeqScore',\n",
    "                                              'ATAC overlap with peak proportion'])['doubleZscore'].mean().reset_index()\n",
    "test_df_unique = test_df_annotated.groupby(['Target Gene Symbol','sgRNA \\'Cut\\' Site TSS Offset','chromosome','Strand of Target',\n",
    "                                              'sgRNA \\'Cut\\' Position','sgRNA Context Sequence','Orientation','rs3ChenSeqScore',\n",
    "                                              'ATAC overlap with peak proportion'])['doubleZscore'].mean().reset_index()\n",
    "train_df_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b893312c-1eec-4c05-9577-2e677f449be8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        chr17:+:36947955\n",
       "1        chr17:+:36947970\n",
       "2        chr17:+:36947973\n",
       "3        chr17:+:36947983\n",
       "4        chr17:+:36947996\n",
       "               ...       \n",
       "95313     chr1:-:85707486\n",
       "95314     chr1:-:85707485\n",
       "95315     chr1:-:85707409\n",
       "95316     chr1:-:85707408\n",
       "95317     chr1:-:85707402\n",
       "Length: 95318, dtype: object"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df_unique['chromosome_crispick_input'] = train_df_unique['chromosome']\n",
    "train_df_unique['chromosome_crispick_input'] = train_df_unique['chromosome_crispick_input'].replace('chr23','chrX')\n",
    "train_df_unique['chromosome_crispick_input'] = train_df_unique['chromosome_crispick_input'].replace('chr24','chrY')\n",
    "\n",
    "coordinates = (train_df_unique['chromosome_crispick_input'] + ':' +\n",
    "              train_df_unique['Strand of Target'] + ':' +\n",
    "              train_df_unique['sgRNA \\'Cut\\' Position'].astype(int).astype(str))\n",
    "coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b9105697-60db-45fc-baa6-0ea7a9d32a4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_df_unique['coordinates'] = coordinates\n",
    "coordinates.drop_duplicates()[0:50000].to_csv('../../Data/modelTraining/rs3i_training_coordinates_part1.txt',\n",
    "                                     sep = '\\n',header=None,index=False)\n",
    "coordinates.drop_duplicates()[50000:].to_csv('../../Data/modelTraining/rs3i_training_coordinates_part2.txt',\n",
    "                                     sep = '\\n',header=None,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a352b9aa-fb85-4107-9e43-4ebc3f52e58d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        chr2:+:215311137\n",
       "1        chr2:+:215311141\n",
       "2        chr2:+:215311644\n",
       "3        chr2:+:215311648\n",
       "4        chr2:+:215311671\n",
       "               ...       \n",
       "23414     chr1:-:85707486\n",
       "23415     chr1:-:85707485\n",
       "23416     chr1:-:85707409\n",
       "23417     chr1:-:85707408\n",
       "23418     chr1:-:85707402\n",
       "Length: 23419, dtype: object"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df_unique['chromosome_crispick_input'] = test_df_unique['chromosome']\n",
    "test_df_unique['chromosome_crispick_input'] = test_df_unique['chromosome_crispick_input'].replace('chr23','chrX')\n",
    "test_df_unique['chromosome_crispick_input'] = test_df_unique['chromosome_crispick_input'].replace('chr24','chrY')\n",
    "\n",
    "coordinates = (test_df_unique['chromosome_crispick_input'] + ':' +\n",
    "              test_df_unique['Strand of Target'] + ':' +\n",
    "              test_df_unique['sgRNA \\'Cut\\' Position'].astype(int).astype(str))\n",
    "coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f3617850-9e52-4a6f-8588-57c4ee404eb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_df_unique['coordinates'] = coordinates\n",
    "coordinates.drop_duplicates().to_csv('../../Data/modelTraining/rs3i_testing_coordinates.txt',\n",
    "                                     sep = '\\n',header=None,index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81fa9ef5-9bed-4462-9438-2e6737728cfb",
   "metadata": {},
   "source": [
    "#### Run CRISPick with following inputs:\n",
    "\n",
    "Reference genome: Ensembl v.114\\\n",
    "Mechanism: CRISPRi\\\n",
    "Enzyme: SpyoCas9\\\n",
    "On-Target Scorer: RS3i-Chen2013+ATAC\\\n",
    "Off-Target Scorer: Seed Score v1\\\n",
    "Targets: upload file with coordinates\\\n",
    "Quota: 1\\\n",
    "Click report unpicked sequences\\\n",
    "TSS range: [-20,20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b7b65f7e-128c-4d48-8ee9-db408f496c8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/mj/lhpd16ys0cx2wkbhm7s0nfcm0000gp/T/ipykernel_69646/3054664851.py:2: DtypeWarning: Columns (24,25) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  crispick_train_part1 = pd.read_table(zip_file.open('rs3i_training_designs_part1.txt'))\n",
      "/var/folders/mj/lhpd16ys0cx2wkbhm7s0nfcm0000gp/T/ipykernel_69646/3054664851.py:4: DtypeWarning: Columns (24,25) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  crispick_train_part2 = pd.read_table(zip_file.open('rs3i_training_designs_part2.txt'))\n"
     ]
    }
   ],
   "source": [
    "zip_file = ZipFile('../../Data/modelTraining/rs3i_training_designs_part1.txt.zip')\n",
    "crispick_train_part1 = pd.read_table(zip_file.open('rs3i_training_designs_part1.txt'))\n",
    "zip_file = ZipFile('../../Data/modelTraining/rs3i_training_designs_part2.txt.zip')\n",
    "crispick_train_part2 = pd.read_table(zip_file.open('rs3i_training_designs_part2.txt'))\n",
    "crispick_test = pd.read_table('../../Data/modelTraining/rs3i_testing_designs.txt')\n",
    "crispick = pd.concat([crispick_train_part1,crispick_train_part2,crispick_test])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "200a3ce4-6579-4e8b-b1d7-4423b8a2dc14",
   "metadata": {},
   "outputs": [],
   "source": [
    "#left merge to see if there are any training data that do not get an atac-seq score\n",
    "train_df_with_atac = train_df_unique.merge(crispick[['Input','sgRNA Context Sequence','ATAC-seq Score']].drop_duplicates(),\n",
    "                                        left_on = ['coordinates','sgRNA Context Sequence'],\n",
    "                                        right_on = ['Input','sgRNA Context Sequence'],\n",
    "                                        how = 'left')\n",
    "test_df_with_atac = test_df_unique.merge(crispick[['Input','sgRNA Context Sequence','ATAC-seq Score']].drop_duplicates(),\n",
    "                                        left_on = ['coordinates','sgRNA Context Sequence'],\n",
    "                                        right_on = ['Input','sgRNA Context Sequence'],\n",
    "                                        how = 'left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "bfa016eb-daef-4f18-b76c-bfa722a7bd57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check that all guides got an atac-seq score\n",
    "train_df_with_atac['ATAC-seq Score'].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "464d8864-9fa0-4f22-bc64-376a0e109679",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df_with_atac['ATAC-seq Score'].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f2e82899-f5d0-43c6-941f-660ea0a88fab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9976289892779958"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#see that most of atac scores agree\n",
    "(train_df_with_atac['ATAC-seq Score'] == round(train_df_with_atac['ATAC overlap with peak proportion'],4)).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f1f5491f-2926-498f-b10a-1e861c9ab080",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9976941799393655"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(test_df_with_atac['ATAC-seq Score'] == round(test_df_with_atac['ATAC overlap with peak proportion'],4)).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3d1f1f5f-cf4b-477d-9839-2ec4daf8841b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_column_name(name):\n",
    "    return (\n",
    "        str(name)\n",
    "        .replace('[', '_')\n",
    "        .replace(']', '_')\n",
    "        .replace('<', '_')\n",
    "        .replace('>', '_')\n",
    "        .replace(' ', '_')\n",
    "        .replace(',', '_')\n",
    "        .replace('(', '_')\n",
    "        .replace(')', '_')\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "30088cb9-7cec-49d9-9d64-4e6c4bc6944a",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_df_with_atac.columns = [clean_column_name(col) for col in train_df_with_atac.columns]\n",
    "test_df_with_atac.columns = [clean_column_name(col) for col in test_df_with_atac.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "cc361ea5-e875-45e5-9d14-9ba36d1f1635",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Target_Gene_Symbol</th>\n",
       "      <th>sgRNA_'Cut'_Site_TSS_Offset</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>Strand_of_Target</th>\n",
       "      <th>sgRNA_'Cut'_Position</th>\n",
       "      <th>sgRNA_Context_Sequence</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>rs3ChenSeqScore</th>\n",
       "      <th>ATAC_overlap_with_peak_proportion</th>\n",
       "      <th>doubleZscore</th>\n",
       "      <th>chromosome_crispick_input</th>\n",
       "      <th>coordinates</th>\n",
       "      <th>Input</th>\n",
       "      <th>ATAC-seq_Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AATF</td>\n",
       "      <td>-1000.0</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>36947955.0</td>\n",
       "      <td>GTTTAGAGGGGTCTAGGTCTCTACAGGCAA</td>\n",
       "      <td>++</td>\n",
       "      <td>0.782222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.055701</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17:+:36947955</td>\n",
       "      <td>chr17:+:36947955</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AATF</td>\n",
       "      <td>-985.0</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>36947970.0</td>\n",
       "      <td>GTCTATTCTTATGGTATCTGTTGGAGGTTG</td>\n",
       "      <td>+-</td>\n",
       "      <td>-0.267594</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.080429</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17:+:36947970</td>\n",
       "      <td>chr17:+:36947970</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AATF</td>\n",
       "      <td>-982.0</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>36947973.0</td>\n",
       "      <td>GTGGTCTATTCTTATGGTATCTGTTGGAGG</td>\n",
       "      <td>+-</td>\n",
       "      <td>0.110804</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.836591</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17:+:36947973</td>\n",
       "      <td>chr17:+:36947973</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AATF</td>\n",
       "      <td>-972.0</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>36947983.0</td>\n",
       "      <td>TGACTACAAAGTGGTCTATTCTTATGGTAT</td>\n",
       "      <td>+-</td>\n",
       "      <td>-0.103015</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.696359</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17:+:36947983</td>\n",
       "      <td>chr17:+:36947983</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AATF</td>\n",
       "      <td>-959.0</td>\n",
       "      <td>chr17</td>\n",
       "      <td>+</td>\n",
       "      <td>36947996.0</td>\n",
       "      <td>GGGATGATAAGTGTGACTACAAAGTGGTCT</td>\n",
       "      <td>+-</td>\n",
       "      <td>0.506514</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.085971</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17:+:36947996</td>\n",
       "      <td>chr17:+:36947996</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95313</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>948.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>85707486.0</td>\n",
       "      <td>GGAGTCAGACGGCGTTCGCGGGGGTGGGTT</td>\n",
       "      <td>--</td>\n",
       "      <td>-1.255058</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>-0.840383</td>\n",
       "      <td>chr1</td>\n",
       "      <td>chr1:-:85707486</td>\n",
       "      <td>chr1:-:85707486</td>\n",
       "      <td>0.1429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95314</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>949.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>85707485.0</td>\n",
       "      <td>GAGTCAGACGGCGTTCGCGGGGGTGGGTTG</td>\n",
       "      <td>--</td>\n",
       "      <td>-0.381792</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>-0.637729</td>\n",
       "      <td>chr1</td>\n",
       "      <td>chr1:-:85707485</td>\n",
       "      <td>chr1:-:85707485</td>\n",
       "      <td>0.1429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95315</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>1025.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>85707409.0</td>\n",
       "      <td>ATACATGAAATTAGTAGTGTCCAATGGGTA</td>\n",
       "      <td>--</td>\n",
       "      <td>-0.244887</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.527092</td>\n",
       "      <td>chr1</td>\n",
       "      <td>chr1:-:85707409</td>\n",
       "      <td>chr1:-:85707409</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95316</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>1026.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>85707408.0</td>\n",
       "      <td>TACATGAAATTAGTAGTGTCCAATGGGTAA</td>\n",
       "      <td>--</td>\n",
       "      <td>0.080581</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.173052</td>\n",
       "      <td>chr1</td>\n",
       "      <td>chr1:-:85707408</td>\n",
       "      <td>chr1:-:85707408</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95317</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>1032.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>85707402.0</td>\n",
       "      <td>AAATTAGTAGTGTCCAATGGGTAATGGTCA</td>\n",
       "      <td>--</td>\n",
       "      <td>0.326828</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.626674</td>\n",
       "      <td>chr1</td>\n",
       "      <td>chr1:-:85707402</td>\n",
       "      <td>chr1:-:85707402</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>95318 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Target_Gene_Symbol  sgRNA_'Cut'_Site_TSS_Offset chromosome  \\\n",
       "0                   AATF                      -1000.0      chr17   \n",
       "1                   AATF                       -985.0      chr17   \n",
       "2                   AATF                       -982.0      chr17   \n",
       "3                   AATF                       -972.0      chr17   \n",
       "4                   AATF                       -959.0      chr17   \n",
       "...                  ...                          ...        ...   \n",
       "95313             ZNHIT6                        948.0       chr1   \n",
       "95314             ZNHIT6                        949.0       chr1   \n",
       "95315             ZNHIT6                       1025.0       chr1   \n",
       "95316             ZNHIT6                       1026.0       chr1   \n",
       "95317             ZNHIT6                       1032.0       chr1   \n",
       "\n",
       "      Strand_of_Target  sgRNA_'Cut'_Position          sgRNA_Context_Sequence  \\\n",
       "0                    +            36947955.0  GTTTAGAGGGGTCTAGGTCTCTACAGGCAA   \n",
       "1                    +            36947970.0  GTCTATTCTTATGGTATCTGTTGGAGGTTG   \n",
       "2                    +            36947973.0  GTGGTCTATTCTTATGGTATCTGTTGGAGG   \n",
       "3                    +            36947983.0  TGACTACAAAGTGGTCTATTCTTATGGTAT   \n",
       "4                    +            36947996.0  GGGATGATAAGTGTGACTACAAAGTGGTCT   \n",
       "...                ...                   ...                             ...   \n",
       "95313                -            85707486.0  GGAGTCAGACGGCGTTCGCGGGGGTGGGTT   \n",
       "95314                -            85707485.0  GAGTCAGACGGCGTTCGCGGGGGTGGGTTG   \n",
       "95315                -            85707409.0  ATACATGAAATTAGTAGTGTCCAATGGGTA   \n",
       "95316                -            85707408.0  TACATGAAATTAGTAGTGTCCAATGGGTAA   \n",
       "95317                -            85707402.0  AAATTAGTAGTGTCCAATGGGTAATGGTCA   \n",
       "\n",
       "      Orientation  rs3ChenSeqScore  ATAC_overlap_with_peak_proportion  \\\n",
       "0              ++         0.782222                           0.000000   \n",
       "1              +-        -0.267594                           0.000000   \n",
       "2              +-         0.110804                           0.000000   \n",
       "3              +-        -0.103015                           0.000000   \n",
       "4              +-         0.506514                           0.000000   \n",
       "...           ...              ...                                ...   \n",
       "95313          --        -1.255058                           0.142857   \n",
       "95314          --        -0.381792                           0.142857   \n",
       "95315          --        -0.244887                           0.000000   \n",
       "95316          --         0.080581                           0.000000   \n",
       "95317          --         0.326828                           0.000000   \n",
       "\n",
       "       doubleZscore chromosome_crispick_input       coordinates  \\\n",
       "0         -1.055701                     chr17  chr17:+:36947955   \n",
       "1         -1.080429                     chr17  chr17:+:36947970   \n",
       "2         -0.836591                     chr17  chr17:+:36947973   \n",
       "3         -0.696359                     chr17  chr17:+:36947983   \n",
       "4          1.085971                     chr17  chr17:+:36947996   \n",
       "...             ...                       ...               ...   \n",
       "95313     -0.840383                      chr1   chr1:-:85707486   \n",
       "95314     -0.637729                      chr1   chr1:-:85707485   \n",
       "95315     -0.527092                      chr1   chr1:-:85707409   \n",
       "95316     -0.173052                      chr1   chr1:-:85707408   \n",
       "95317     -0.626674                      chr1   chr1:-:85707402   \n",
       "\n",
       "                  Input  ATAC-seq_Score  \n",
       "0      chr17:+:36947955          0.0000  \n",
       "1      chr17:+:36947970          0.0000  \n",
       "2      chr17:+:36947973          0.0000  \n",
       "3      chr17:+:36947983          0.0000  \n",
       "4      chr17:+:36947996          0.0000  \n",
       "...                 ...             ...  \n",
       "95313   chr1:-:85707486          0.1429  \n",
       "95314   chr1:-:85707485          0.1429  \n",
       "95315   chr1:-:85707409          0.0000  \n",
       "95316   chr1:-:85707408          0.0000  \n",
       "95317   chr1:-:85707402          0.0000  \n",
       "\n",
       "[95318 rows x 14 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df_with_atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "dc35823f-5cc7-4f23-9596-54d15b914418",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Target_Gene_Symbol</th>\n",
       "      <th>sgRNA_'Cut'_Site_TSS_Offset</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>Strand_of_Target</th>\n",
       "      <th>sgRNA_'Cut'_Position</th>\n",
       "      <th>sgRNA_Context_Sequence</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>rs3ChenSeqScore</th>\n",
       "      <th>ATAC_overlap_with_peak_proportion</th>\n",
       "      <th>doubleZscore</th>\n",
       "      <th>chromosome_crispick_input</th>\n",
       "      <th>coordinates</th>\n",
       "      <th>Input</th>\n",
       "      <th>ATAC-seq_Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ATIC</td>\n",
       "      <td>-923.0</td>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>215311137.0</td>\n",
       "      <td>AAAATTAGACGGGTGTGGAGGCAGAGGCAG</td>\n",
       "      <td>++</td>\n",
       "      <td>0.136541</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>-0.199624</td>\n",
       "      <td>chr2</td>\n",
       "      <td>chr2:+:215311137</td>\n",
       "      <td>chr2:+:215311137</td>\n",
       "      <td>0.1429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ATIC</td>\n",
       "      <td>-919.0</td>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>215311141.0</td>\n",
       "      <td>TTAGACGGGTGTGGAGGCAGAGGCAGGAGA</td>\n",
       "      <td>++</td>\n",
       "      <td>-0.438222</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>-0.419226</td>\n",
       "      <td>chr2</td>\n",
       "      <td>chr2:+:215311141</td>\n",
       "      <td>chr2:+:215311141</td>\n",
       "      <td>0.1429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATIC</td>\n",
       "      <td>-416.0</td>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>215311644.0</td>\n",
       "      <td>CAAAAAAACAACAAACAAAAAGGGAGGATG</td>\n",
       "      <td>++</td>\n",
       "      <td>-0.252245</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.156149</td>\n",
       "      <td>chr2</td>\n",
       "      <td>chr2:+:215311644</td>\n",
       "      <td>chr2:+:215311644</td>\n",
       "      <td>0.2857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ATIC</td>\n",
       "      <td>-412.0</td>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>215311648.0</td>\n",
       "      <td>AAAACAACAAACAAAAAGGGAGGATGGTAG</td>\n",
       "      <td>++</td>\n",
       "      <td>-0.385688</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>-0.152352</td>\n",
       "      <td>chr2</td>\n",
       "      <td>chr2:+:215311648</td>\n",
       "      <td>chr2:+:215311648</td>\n",
       "      <td>0.2857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATIC</td>\n",
       "      <td>-389.0</td>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>215311671.0</td>\n",
       "      <td>ATAGCGGCAAAGCTGGAAAGTATCTGGACA</td>\n",
       "      <td>+-</td>\n",
       "      <td>-0.517484</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>-0.423902</td>\n",
       "      <td>chr2</td>\n",
       "      <td>chr2:+:215311671</td>\n",
       "      <td>chr2:+:215311671</td>\n",
       "      <td>0.2857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23414</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>948.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>85707486.0</td>\n",
       "      <td>GGAGTCAGACGGCGTTCGCGGGGGTGGGTT</td>\n",
       "      <td>--</td>\n",
       "      <td>-1.255058</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>-0.840383</td>\n",
       "      <td>chr1</td>\n",
       "      <td>chr1:-:85707486</td>\n",
       "      <td>chr1:-:85707486</td>\n",
       "      <td>0.1429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23415</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>949.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>85707485.0</td>\n",
       "      <td>GAGTCAGACGGCGTTCGCGGGGGTGGGTTG</td>\n",
       "      <td>--</td>\n",
       "      <td>-0.381792</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>-0.637729</td>\n",
       "      <td>chr1</td>\n",
       "      <td>chr1:-:85707485</td>\n",
       "      <td>chr1:-:85707485</td>\n",
       "      <td>0.1429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23416</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>1025.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>85707409.0</td>\n",
       "      <td>ATACATGAAATTAGTAGTGTCCAATGGGTA</td>\n",
       "      <td>--</td>\n",
       "      <td>-0.244887</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.527092</td>\n",
       "      <td>chr1</td>\n",
       "      <td>chr1:-:85707409</td>\n",
       "      <td>chr1:-:85707409</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23417</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>1026.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>85707408.0</td>\n",
       "      <td>TACATGAAATTAGTAGTGTCCAATGGGTAA</td>\n",
       "      <td>--</td>\n",
       "      <td>0.080581</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.173052</td>\n",
       "      <td>chr1</td>\n",
       "      <td>chr1:-:85707408</td>\n",
       "      <td>chr1:-:85707408</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23418</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>1032.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>85707402.0</td>\n",
       "      <td>AAATTAGTAGTGTCCAATGGGTAATGGTCA</td>\n",
       "      <td>--</td>\n",
       "      <td>0.326828</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.626674</td>\n",
       "      <td>chr1</td>\n",
       "      <td>chr1:-:85707402</td>\n",
       "      <td>chr1:-:85707402</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23419 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Target_Gene_Symbol  sgRNA_'Cut'_Site_TSS_Offset chromosome  \\\n",
       "0                   ATIC                       -923.0       chr2   \n",
       "1                   ATIC                       -919.0       chr2   \n",
       "2                   ATIC                       -416.0       chr2   \n",
       "3                   ATIC                       -412.0       chr2   \n",
       "4                   ATIC                       -389.0       chr2   \n",
       "...                  ...                          ...        ...   \n",
       "23414             ZNHIT6                        948.0       chr1   \n",
       "23415             ZNHIT6                        949.0       chr1   \n",
       "23416             ZNHIT6                       1025.0       chr1   \n",
       "23417             ZNHIT6                       1026.0       chr1   \n",
       "23418             ZNHIT6                       1032.0       chr1   \n",
       "\n",
       "      Strand_of_Target  sgRNA_'Cut'_Position          sgRNA_Context_Sequence  \\\n",
       "0                    +           215311137.0  AAAATTAGACGGGTGTGGAGGCAGAGGCAG   \n",
       "1                    +           215311141.0  TTAGACGGGTGTGGAGGCAGAGGCAGGAGA   \n",
       "2                    +           215311644.0  CAAAAAAACAACAAACAAAAAGGGAGGATG   \n",
       "3                    +           215311648.0  AAAACAACAAACAAAAAGGGAGGATGGTAG   \n",
       "4                    +           215311671.0  ATAGCGGCAAAGCTGGAAAGTATCTGGACA   \n",
       "...                ...                   ...                             ...   \n",
       "23414                -            85707486.0  GGAGTCAGACGGCGTTCGCGGGGGTGGGTT   \n",
       "23415                -            85707485.0  GAGTCAGACGGCGTTCGCGGGGGTGGGTTG   \n",
       "23416                -            85707409.0  ATACATGAAATTAGTAGTGTCCAATGGGTA   \n",
       "23417                -            85707408.0  TACATGAAATTAGTAGTGTCCAATGGGTAA   \n",
       "23418                -            85707402.0  AAATTAGTAGTGTCCAATGGGTAATGGTCA   \n",
       "\n",
       "      Orientation  rs3ChenSeqScore  ATAC_overlap_with_peak_proportion  \\\n",
       "0              ++         0.136541                           0.142857   \n",
       "1              ++        -0.438222                           0.142857   \n",
       "2              ++        -0.252245                           0.285714   \n",
       "3              ++        -0.385688                           0.285714   \n",
       "4              +-        -0.517484                           0.285714   \n",
       "...           ...              ...                                ...   \n",
       "23414          --        -1.255058                           0.142857   \n",
       "23415          --        -0.381792                           0.142857   \n",
       "23416          --        -0.244887                           0.000000   \n",
       "23417          --         0.080581                           0.000000   \n",
       "23418          --         0.326828                           0.000000   \n",
       "\n",
       "       doubleZscore chromosome_crispick_input       coordinates  \\\n",
       "0         -0.199624                      chr2  chr2:+:215311137   \n",
       "1         -0.419226                      chr2  chr2:+:215311141   \n",
       "2          0.156149                      chr2  chr2:+:215311644   \n",
       "3         -0.152352                      chr2  chr2:+:215311648   \n",
       "4         -0.423902                      chr2  chr2:+:215311671   \n",
       "...             ...                       ...               ...   \n",
       "23414     -0.840383                      chr1   chr1:-:85707486   \n",
       "23415     -0.637729                      chr1   chr1:-:85707485   \n",
       "23416     -0.527092                      chr1   chr1:-:85707409   \n",
       "23417     -0.173052                      chr1   chr1:-:85707408   \n",
       "23418     -0.626674                      chr1   chr1:-:85707402   \n",
       "\n",
       "                  Input  ATAC-seq_Score  \n",
       "0      chr2:+:215311137          0.1429  \n",
       "1      chr2:+:215311141          0.1429  \n",
       "2      chr2:+:215311644          0.2857  \n",
       "3      chr2:+:215311648          0.2857  \n",
       "4      chr2:+:215311671          0.2857  \n",
       "...                 ...             ...  \n",
       "23414   chr1:-:85707486          0.1429  \n",
       "23415   chr1:-:85707485          0.1429  \n",
       "23416   chr1:-:85707409          0.0000  \n",
       "23417   chr1:-:85707408          0.0000  \n",
       "23418   chr1:-:85707402          0.0000  \n",
       "\n",
       "[23419 rows x 14 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df_with_atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "9b738932-dc52-4323-9493-b73f55db089e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/mj/lhpd16ys0cx2wkbhm7s0nfcm0000gp/T/ipykernel_69646/2139439996.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  X_train.loc[:,'rs3ChenSeqScore'] = round(X_train['rs3ChenSeqScore'],4)\n",
      "/var/folders/mj/lhpd16ys0cx2wkbhm7s0nfcm0000gp/T/ipykernel_69646/2139439996.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  X_train.loc[:,'sgRNA_\\'Cut\\'_Site_TSS_Offset'] = X_train['sgRNA_\\'Cut\\'_Site_TSS_Offset'].astype(int)\n",
      "/var/folders/mj/lhpd16ys0cx2wkbhm7s0nfcm0000gp/T/ipykernel_69646/2139439996.py:3: DeprecationWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`\n",
      "  X_train.loc[:,'sgRNA_\\'Cut\\'_Site_TSS_Offset'] = X_train['sgRNA_\\'Cut\\'_Site_TSS_Offset'].astype(int)\n",
      "/var/folders/mj/lhpd16ys0cx2wkbhm7s0nfcm0000gp/T/ipykernel_69646/2139439996.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  X_train['ATAC_overlap_with_peak'] = pd.cut(train_df_with_atac['ATAC-seq_Score'],bins=[0,0.5,1,np.inf],right=False)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Target_Gene_Symbol</th>\n",
       "      <th>sgRNA_Context_Sequence</th>\n",
       "      <th>sgRNA_'Cut'_Site_TSS_Offset</th>\n",
       "      <th>rs3ChenSeqScore</th>\n",
       "      <th>doubleZscore</th>\n",
       "      <th>ATAC_overlap_with_peak__0.0__0.5_</th>\n",
       "      <th>ATAC_overlap_with_peak__0.5__1.0_</th>\n",
       "      <th>ATAC_overlap_with_peak__1.0__inf_</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AATF</td>\n",
       "      <td>GTTTAGAGGGGTCTAGGTCTCTACAGGCAA</td>\n",
       "      <td>-1000</td>\n",
       "      <td>0.7822</td>\n",
       "      <td>-1.055701</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AATF</td>\n",
       "      <td>GTCTATTCTTATGGTATCTGTTGGAGGTTG</td>\n",
       "      <td>-985</td>\n",
       "      <td>-0.2676</td>\n",
       "      <td>-1.080429</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AATF</td>\n",
       "      <td>GTGGTCTATTCTTATGGTATCTGTTGGAGG</td>\n",
       "      <td>-982</td>\n",
       "      <td>0.1108</td>\n",
       "      <td>-0.836591</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AATF</td>\n",
       "      <td>TGACTACAAAGTGGTCTATTCTTATGGTAT</td>\n",
       "      <td>-972</td>\n",
       "      <td>-0.1030</td>\n",
       "      <td>-0.696359</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AATF</td>\n",
       "      <td>GGGATGATAAGTGTGACTACAAAGTGGTCT</td>\n",
       "      <td>-959</td>\n",
       "      <td>0.5065</td>\n",
       "      <td>1.085971</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95313</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>GGAGTCAGACGGCGTTCGCGGGGGTGGGTT</td>\n",
       "      <td>948</td>\n",
       "      <td>-1.2551</td>\n",
       "      <td>-0.840383</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95314</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>GAGTCAGACGGCGTTCGCGGGGGTGGGTTG</td>\n",
       "      <td>949</td>\n",
       "      <td>-0.3818</td>\n",
       "      <td>-0.637729</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95315</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>ATACATGAAATTAGTAGTGTCCAATGGGTA</td>\n",
       "      <td>1025</td>\n",
       "      <td>-0.2449</td>\n",
       "      <td>-0.527092</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95316</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>TACATGAAATTAGTAGTGTCCAATGGGTAA</td>\n",
       "      <td>1026</td>\n",
       "      <td>0.0806</td>\n",
       "      <td>-0.173052</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95317</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>AAATTAGTAGTGTCCAATGGGTAATGGTCA</td>\n",
       "      <td>1032</td>\n",
       "      <td>0.3268</td>\n",
       "      <td>-0.626674</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>95318 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Target_Gene_Symbol          sgRNA_Context_Sequence  \\\n",
       "0                   AATF  GTTTAGAGGGGTCTAGGTCTCTACAGGCAA   \n",
       "1                   AATF  GTCTATTCTTATGGTATCTGTTGGAGGTTG   \n",
       "2                   AATF  GTGGTCTATTCTTATGGTATCTGTTGGAGG   \n",
       "3                   AATF  TGACTACAAAGTGGTCTATTCTTATGGTAT   \n",
       "4                   AATF  GGGATGATAAGTGTGACTACAAAGTGGTCT   \n",
       "...                  ...                             ...   \n",
       "95313             ZNHIT6  GGAGTCAGACGGCGTTCGCGGGGGTGGGTT   \n",
       "95314             ZNHIT6  GAGTCAGACGGCGTTCGCGGGGGTGGGTTG   \n",
       "95315             ZNHIT6  ATACATGAAATTAGTAGTGTCCAATGGGTA   \n",
       "95316             ZNHIT6  TACATGAAATTAGTAGTGTCCAATGGGTAA   \n",
       "95317             ZNHIT6  AAATTAGTAGTGTCCAATGGGTAATGGTCA   \n",
       "\n",
       "       sgRNA_'Cut'_Site_TSS_Offset  rs3ChenSeqScore  doubleZscore  \\\n",
       "0                            -1000           0.7822     -1.055701   \n",
       "1                             -985          -0.2676     -1.080429   \n",
       "2                             -982           0.1108     -0.836591   \n",
       "3                             -972          -0.1030     -0.696359   \n",
       "4                             -959           0.5065      1.085971   \n",
       "...                            ...              ...           ...   \n",
       "95313                          948          -1.2551     -0.840383   \n",
       "95314                          949          -0.3818     -0.637729   \n",
       "95315                         1025          -0.2449     -0.527092   \n",
       "95316                         1026           0.0806     -0.173052   \n",
       "95317                         1032           0.3268     -0.626674   \n",
       "\n",
       "       ATAC_overlap_with_peak__0.0__0.5_  ATAC_overlap_with_peak__0.5__1.0_  \\\n",
       "0                                      1                                  0   \n",
       "1                                      1                                  0   \n",
       "2                                      1                                  0   \n",
       "3                                      1                                  0   \n",
       "4                                      1                                  0   \n",
       "...                                  ...                                ...   \n",
       "95313                                  1                                  0   \n",
       "95314                                  1                                  0   \n",
       "95315                                  1                                  0   \n",
       "95316                                  1                                  0   \n",
       "95317                                  1                                  0   \n",
       "\n",
       "       ATAC_overlap_with_peak__1.0__inf_  \n",
       "0                                      0  \n",
       "1                                      0  \n",
       "2                                      0  \n",
       "3                                      0  \n",
       "4                                      0  \n",
       "...                                  ...  \n",
       "95313                                  0  \n",
       "95314                                  0  \n",
       "95315                                  0  \n",
       "95316                                  0  \n",
       "95317                                  0  \n",
       "\n",
       "[95318 rows x 8 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_train = train_df_with_atac[['Target_Gene_Symbol','sgRNA_Context_Sequence','sgRNA_\\'Cut\\'_Site_TSS_Offset', 'rs3ChenSeqScore','doubleZscore']]\n",
    "X_train.loc[:,'rs3ChenSeqScore'] = round(X_train['rs3ChenSeqScore'],4)\n",
    "X_train.loc[:,'sgRNA_\\'Cut\\'_Site_TSS_Offset'] = X_train['sgRNA_\\'Cut\\'_Site_TSS_Offset'].astype(int)\n",
    "X_train['ATAC_overlap_with_peak'] = pd.cut(train_df_with_atac['ATAC-seq_Score'],bins=[0,0.5,1,np.inf],right=False)\n",
    "\n",
    "# one-hot encoding\n",
    "X_train_encoded = pd.get_dummies(X_train, columns=['ATAC_overlap_with_peak'])\n",
    "\n",
    "bool_cols = X_train_encoded.select_dtypes(include=['bool']).columns.tolist()\n",
    "X_train_encoded[bool_cols] = X_train_encoded[bool_cols].astype(int)\n",
    "\n",
    "def clean_column_name(name):\n",
    "    return (\n",
    "        str(name)\n",
    "        .replace('[', '_')\n",
    "        .replace(']', '_')\n",
    "        .replace('<', '_')\n",
    "        .replace('>', '_')\n",
    "        .replace(' ', '_')\n",
    "        .replace(',', '_')\n",
    "        .replace('(', '_')\n",
    "        .replace(')', '_')\n",
    "    )\n",
    "\n",
    "X_train_encoded.columns = [clean_column_name(col) for col in X_train_encoded.columns]\n",
    "X_train_encoded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3724f3bb-03b5-4489-9948-8770e20d4434",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/mj/lhpd16ys0cx2wkbhm7s0nfcm0000gp/T/ipykernel_69646/3623459053.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  X_test.loc[:,'rs3ChenSeqScore'] = round(X_test['rs3ChenSeqScore'],4)\n",
      "/var/folders/mj/lhpd16ys0cx2wkbhm7s0nfcm0000gp/T/ipykernel_69646/3623459053.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  X_test.loc[:,'sgRNA_\\'Cut\\'_Site_TSS_Offset'] = X_test['sgRNA_\\'Cut\\'_Site_TSS_Offset'].astype(int)\n",
      "/var/folders/mj/lhpd16ys0cx2wkbhm7s0nfcm0000gp/T/ipykernel_69646/3623459053.py:3: DeprecationWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`\n",
      "  X_test.loc[:,'sgRNA_\\'Cut\\'_Site_TSS_Offset'] = X_test['sgRNA_\\'Cut\\'_Site_TSS_Offset'].astype(int)\n",
      "/var/folders/mj/lhpd16ys0cx2wkbhm7s0nfcm0000gp/T/ipykernel_69646/3623459053.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  X_test['ATAC_overlap_with_peak'] = pd.cut(test_df_with_atac['ATAC-seq_Score'],bins=[0,0.5,1,np.inf],right=False)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Target_Gene_Symbol</th>\n",
       "      <th>sgRNA_Context_Sequence</th>\n",
       "      <th>sgRNA_'Cut'_Site_TSS_Offset</th>\n",
       "      <th>rs3ChenSeqScore</th>\n",
       "      <th>doubleZscore</th>\n",
       "      <th>ATAC_overlap_with_peak__0.0__0.5_</th>\n",
       "      <th>ATAC_overlap_with_peak__0.5__1.0_</th>\n",
       "      <th>ATAC_overlap_with_peak__1.0__inf_</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ATIC</td>\n",
       "      <td>AAAATTAGACGGGTGTGGAGGCAGAGGCAG</td>\n",
       "      <td>-923</td>\n",
       "      <td>0.1365</td>\n",
       "      <td>-0.199624</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ATIC</td>\n",
       "      <td>TTAGACGGGTGTGGAGGCAGAGGCAGGAGA</td>\n",
       "      <td>-919</td>\n",
       "      <td>-0.4382</td>\n",
       "      <td>-0.419226</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATIC</td>\n",
       "      <td>CAAAAAAACAACAAACAAAAAGGGAGGATG</td>\n",
       "      <td>-416</td>\n",
       "      <td>-0.2522</td>\n",
       "      <td>0.156149</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ATIC</td>\n",
       "      <td>AAAACAACAAACAAAAAGGGAGGATGGTAG</td>\n",
       "      <td>-412</td>\n",
       "      <td>-0.3857</td>\n",
       "      <td>-0.152352</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATIC</td>\n",
       "      <td>ATAGCGGCAAAGCTGGAAAGTATCTGGACA</td>\n",
       "      <td>-389</td>\n",
       "      <td>-0.5175</td>\n",
       "      <td>-0.423902</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23414</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>GGAGTCAGACGGCGTTCGCGGGGGTGGGTT</td>\n",
       "      <td>948</td>\n",
       "      <td>-1.2551</td>\n",
       "      <td>-0.840383</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23415</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>GAGTCAGACGGCGTTCGCGGGGGTGGGTTG</td>\n",
       "      <td>949</td>\n",
       "      <td>-0.3818</td>\n",
       "      <td>-0.637729</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23416</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>ATACATGAAATTAGTAGTGTCCAATGGGTA</td>\n",
       "      <td>1025</td>\n",
       "      <td>-0.2449</td>\n",
       "      <td>-0.527092</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23417</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>TACATGAAATTAGTAGTGTCCAATGGGTAA</td>\n",
       "      <td>1026</td>\n",
       "      <td>0.0806</td>\n",
       "      <td>-0.173052</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23418</th>\n",
       "      <td>ZNHIT6</td>\n",
       "      <td>AAATTAGTAGTGTCCAATGGGTAATGGTCA</td>\n",
       "      <td>1032</td>\n",
       "      <td>0.3268</td>\n",
       "      <td>-0.626674</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23419 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Target_Gene_Symbol          sgRNA_Context_Sequence  \\\n",
       "0                   ATIC  AAAATTAGACGGGTGTGGAGGCAGAGGCAG   \n",
       "1                   ATIC  TTAGACGGGTGTGGAGGCAGAGGCAGGAGA   \n",
       "2                   ATIC  CAAAAAAACAACAAACAAAAAGGGAGGATG   \n",
       "3                   ATIC  AAAACAACAAACAAAAAGGGAGGATGGTAG   \n",
       "4                   ATIC  ATAGCGGCAAAGCTGGAAAGTATCTGGACA   \n",
       "...                  ...                             ...   \n",
       "23414             ZNHIT6  GGAGTCAGACGGCGTTCGCGGGGGTGGGTT   \n",
       "23415             ZNHIT6  GAGTCAGACGGCGTTCGCGGGGGTGGGTTG   \n",
       "23416             ZNHIT6  ATACATGAAATTAGTAGTGTCCAATGGGTA   \n",
       "23417             ZNHIT6  TACATGAAATTAGTAGTGTCCAATGGGTAA   \n",
       "23418             ZNHIT6  AAATTAGTAGTGTCCAATGGGTAATGGTCA   \n",
       "\n",
       "       sgRNA_'Cut'_Site_TSS_Offset  rs3ChenSeqScore  doubleZscore  \\\n",
       "0                             -923           0.1365     -0.199624   \n",
       "1                             -919          -0.4382     -0.419226   \n",
       "2                             -416          -0.2522      0.156149   \n",
       "3                             -412          -0.3857     -0.152352   \n",
       "4                             -389          -0.5175     -0.423902   \n",
       "...                            ...              ...           ...   \n",
       "23414                          948          -1.2551     -0.840383   \n",
       "23415                          949          -0.3818     -0.637729   \n",
       "23416                         1025          -0.2449     -0.527092   \n",
       "23417                         1026           0.0806     -0.173052   \n",
       "23418                         1032           0.3268     -0.626674   \n",
       "\n",
       "       ATAC_overlap_with_peak__0.0__0.5_  ATAC_overlap_with_peak__0.5__1.0_  \\\n",
       "0                                      1                                  0   \n",
       "1                                      1                                  0   \n",
       "2                                      1                                  0   \n",
       "3                                      1                                  0   \n",
       "4                                      1                                  0   \n",
       "...                                  ...                                ...   \n",
       "23414                                  1                                  0   \n",
       "23415                                  1                                  0   \n",
       "23416                                  1                                  0   \n",
       "23417                                  1                                  0   \n",
       "23418                                  1                                  0   \n",
       "\n",
       "       ATAC_overlap_with_peak__1.0__inf_  \n",
       "0                                      0  \n",
       "1                                      0  \n",
       "2                                      0  \n",
       "3                                      0  \n",
       "4                                      0  \n",
       "...                                  ...  \n",
       "23414                                  0  \n",
       "23415                                  0  \n",
       "23416                                  0  \n",
       "23417                                  0  \n",
       "23418                                  0  \n",
       "\n",
       "[23419 rows x 8 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_test = test_df_with_atac[['Target_Gene_Symbol','sgRNA_Context_Sequence','sgRNA_\\'Cut\\'_Site_TSS_Offset', 'rs3ChenSeqScore','doubleZscore']]\n",
    "X_test.loc[:,'rs3ChenSeqScore'] = round(X_test['rs3ChenSeqScore'],4)\n",
    "X_test.loc[:,'sgRNA_\\'Cut\\'_Site_TSS_Offset'] = X_test['sgRNA_\\'Cut\\'_Site_TSS_Offset'].astype(int)\n",
    "X_test['ATAC_overlap_with_peak'] = pd.cut(test_df_with_atac['ATAC-seq_Score'],bins=[0,0.5,1,np.inf],right=False)\n",
    "\n",
    "# one-hot encoding\n",
    "X_test_encoded = pd.get_dummies(X_test, columns=['ATAC_overlap_with_peak'])\n",
    "\n",
    "bool_cols = X_test_encoded.select_dtypes(include=['bool']).columns.tolist()\n",
    "X_test_encoded[bool_cols] = X_test_encoded[bool_cols].astype(int)\n",
    "\n",
    "def clean_column_name(name):\n",
    "    return (\n",
    "        str(name)\n",
    "        .replace('[', '_')\n",
    "        .replace(']', '_')\n",
    "        .replace('<', '_')\n",
    "        .replace('>', '_')\n",
    "        .replace(' ', '_')\n",
    "        .replace(',', '_')\n",
    "        .replace('(', '_')\n",
    "        .replace(')', '_')\n",
    "    )\n",
    "\n",
    "X_test_encoded.columns = [clean_column_name(col) for col in X_test_encoded.columns]\n",
    "X_test_encoded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c4a21645-0cd0-4cd1-bde3-abb59c903b25",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train_encoded.to_csv('../../Data/modelTraining/trainData_rs3i.csv',index=False)\n",
    "X_test_encoded.to_csv('../../Data/modelTraining/testData_rs3i.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb758cba-8b9d-4622-bd85-2baa156c6fbd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "crispri",
   "language": "python",
   "name": "crispri"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
